### DISCUSSION TITLE
Declare class in one line

### DISCUSSION THREAD

**Discussion by lazyloader** [dotnet/csharplang#1659] [Language Ideas] [answered]
Sometimes we use really simple classes to demenstrate simple concepts and they contain only one method or field. Instead of this

```csharp
public class Foo
    {
    public void Bar() => Console.WriteLine("Bar");
    }
```

we can do this

```csharp
public class Foo => public void Bar()  => Console.WriteLine("Bar");
```

And instead of 
```csharp
public class Foo
    {
    public void Bar()  => Console.WriteLine("Bar");
    public int Baz() => Console.WriteLine("Baz");
    }
```

this:
```csharp
public class Foo => public void Bar()  => Console.WriteLine("Bar");, public int Baz()  => Console.WriteLine("Baz");;
```

**Ediit** added implimentations

**Comment by jnm2** [comment]
@lazyloader Not even the first example could be legal C# since non-abstract methods must declare bodies.

**Comment by lazyloader** [comment]
@jnm2 there, added implimentations

**Comment by CyrusNajmabadi** [comment, marked as answer]
Why is:

```c@#
public class Foo => public void Bar() => Console.WriteLine("Bar");
```

better than

```c#
public class Foo { public void Bar() => Console.WriteLine("Bar"); }
```

AFAICT, you save **1** character...

**Comment by CyrusNajmabadi** [comment]
This case is even weirder:

```c#
public class Foo => public void Bar() => Console.WriteLine("Bar");, public int Baz() => Console.WriteLine("Baz");;
```

Because you can already write:

```c#
public class Foo { public void Bar() => Console.WriteLine("Bar"); public int Baz() => Console.WriteLine("Baz"); }
```

Which is one character fewer!  So your 'declare class in one line' proposal actually takes up more space :)

**Comment by jnm2** [comment]
@lazyloader Now that you've added method bodies, it shows that each of the examples is harder to read than the equivalent multiline syntax. And that's just when every method has a very simple body.

Depending on what you're doing, you might be interested in https://github.com/dotnet/csharplang/blob/master/proposals/records.md#records though. I definitely am!

**Comment by lazyloader** [comment]
@CyrusNajmabadi  Couldn't you have made the same arguement for expression body methods?

```csharp
public void Foo()
    {
    Console.WriteLine("Foo");
    }
```

Written as
```csharp
public void Foo() { Console.WriteLine("Foo"); }
```

So why bother with

```csharp
public void Foo() => Console.WriteLine("Foo");
```

It's cleaner and easier to read, and almost no one declares a function with everything on one line unless they're using `=>`. 

Also `=>` can be used in many contexts now to make something one line, so it's inconistent to not let classes or even namespaces be defined in one line using `=>` syntax

**Comment by CyrusNajmabadi** [comment]
>@CyrusNajmabadi Couldn't you have made the same arguement for expression body methods?

Yes.  Bu that wasn't the argument for Expression-Body methods.  Expression-body methods came from the fact that we'd already done ```=>``` for lambdas and then had made them available for properties *where they're very useful*.  i.e. it is a lot nicer to say:

```c#
public int Foo => whatever;
```

Instead of

```c#
public int Foo { get { return whatever; } }
```

Once we already had ```=>``` it made sense to just make it available in more sensible contexts.

--

So the argument was never "let's reduce methods from ```void Foo() { Bar(); }``` to ```void Foo() => Bar();```.  It was "we already have this pattern that we allow in many other places.  Let's just allow it in more.

--

Also, the 'void' version of ```=>``` is far less interesting than the non-void version.  i.e. the much more compelling case is ```string Foo() => Bar();``` over ```string Foo() { return Bar(); }``` which is substantially larger.

**Comment by stuartstein777** [comment]
F# record style declarations would be nice.

**Comment by HaloFour** [comment]
@stuartstein777 

#39
